RavenDB-13940 Adding a configuration option that will allow skipping …

arekpalinski opened the pull request:
…corrupted transactions in journal if they were already synced

arekpalinski committed:
RavenDB-13940 Adding a configuration option that will allow skipping corrupted transactions in journal if they were already synced

ayende commented on src/Raven.Server/Config/Categories/StorageConfiguration.cs:
```diff
@@ -113,5 +113,10 @@ public class StorageConfiguration : ConfigurationCategory
         public bool SkipChecksumValidationOnDatabaseLoading { get; set; }
+
+        [Description("EXPERT: Allows to load a database regardless encountered data integrity errors of already synced transactions in journals during the recovery operation on startup")]
+        [DefaultValue(false)]
```
Shouldn't we always ignore this? Why make this optional?

arekpalinski replied:
I was thinking about enabling it by default, but I don't like to silently ignore this kind of errors, even if that's safe. I prefer a user to make aware of such error. 

I propose that if we detect that a corrupted transaction was corrupted then we'll be thowing still, but we'll include to the exception message info it can be potentially fixed (without any data loss) by enabling this option. What do you think?

ayende replied:
We saw how that turned out with the missing journal. Even with the actual fix in the message , we got a lot of support calls and scared users.
Given that we *know* that this is an okay scenario, no data loss, why force it on the user?

arekpalinski replied:
I think this case is slightly different because I believe that in this scenario users might be running into this due to hardware problem. So we need to let them know about that even if it's safe to continue the recovery operation.

I don't want to do it completely silently. I think this will be very rare BTW

ayende replied:
So let's make this an alert by default? I agree that we want the user to be aware of this.

arekpalinski replied:
Okay. Let's alert them at least.

arekpalinski replied:
Added in https://github.com/ravendb/ravendb/pull/9654/commits/6eb47e553813fbabcb6ccaadf0cdca75d266cbac

ayende commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -54,15 +54,21 @@ public bool ReadOneTransactionToDataFile(StorageEnvironmentOptions options)
                 {
                     if (TryReadAndValidateHeader(options, out current))
                     {
+                        if (CanIgnoreDataIntegrityErrorBecauseTxWasSynced(current, options))
```
Here we need to see if the _next_ transaction is valid, (including the id), and only then agree to jump to it.
Because if the current tx is bad, we can't trust its length either

arekpalinski replied:
Note that here `TryReadAndValidateHeader` returned `true` so we already found a valid transaction. `CanIgnoreDataIntegrityErrorBecauseTxWasSynced` call was to check that we're _still_ in this special mode (corrupted but already synced) and can continue the recovery.

ayende replied:
ok

ayende commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -427,18 +471,34 @@ private bool TryReadAndValidateHeader(StorageEnvironmentOptions options, out Tra
+            // if we have a journal which contains transactions that has been synced and this is the case for current transaction 
+            // then we can continue the recovery regardless encountered errors
+
+            return IsAlreadySyncTransaction(currentTx) && options.IgnoreDataIntegrityErrorsOfAlreadySyncedTransactions;
```
Here, need to (carefully) check that the _next_ tx is valid (proper id and checksum)

arekpalinski replied:
Why? Next transaction can also be corrupted and that's fine I think. Then we're looking for next tx valid header (where you already commented - see my answer there):

https://github.com/ravendb/ravendb/pull/9654/files/2dd60ebbde8ee0da60f0c05559e906da9bd280c6#diff-500338b5995fcc8a61f42b513e64f35eR64-R66

and if we find it and it's _still_ a transaction which was synced then we can skip multiple corrupted transactions.

arekpalinski replied:
For example, we do the recovery, last sync tx is 5:

tx 1 - ok
tx 2 - corrupted
tx 3 - corrupted
tx 4 - valid
tx 5 - valid

We found tx 2 is corrupted then we start looking for next valid transaction. We found tx 4 is valid and was synced so we can safely skip both: tx 2 and tx 3

arekpalinski replied:
Added TODO comment for making sure the following won't pass: https://issues.hibernatingrhinos.com/issue/RavenDB-13940#focus=streamItem-67-40594.0-0

ayende replied:
Okay, make sense

arekpalinski replied:
Test added in https://github.com/ravendb/ravendb/pull/9654/commits/6eb47e553813fbabcb6ccaadf0cdca75d266cbac

ppekrol reviewed (approved):
LGTM

arekpalinski committed:
RavenDB-13940
- Making IgnoreDataIntegrityErrorsOfAlreadySyncedTransactions enabled by default
- Adding alerts if such error is detected
- Added test

arekpalinski committed:
RavenDB-13940 When detecting if we can skip the already sync transaction we need to take into account that a transaction might come from a journal that has been already deleted but we reused a file for next journal.
